File: bridge/project-scan.js
import { existsSync, readFileSync, readdirSync } from 'fs';
import { join, resolve } from 'path';

const ENTRY_FILES = [
  'src/App.tsx',
  'src/App.jsx',
  'src/App.vue',
  'src/main.tsx',
  'src/main.jsx',
  'src/main.ts',
  'src/main.js',
  'src/pages/index.tsx',
  'src/app/page.tsx',
  'app/page.tsx',
  'pages/index.tsx',
  'src/index.tsx',
  'src/index.jsx',
  'index.html'
];

const CONFIG_FILES = [
  'vite.config.ts',
  'vite.config.js',
  'next.config.js',
  'next.config.mjs',
  'webpack.config.js',
  'webpack.config.ts',
  'tsconfig.json',
  'jsconfig.json',
  'tailwind.config.js',
  'postcss.config.js'
];

const KEY_DIRS = [
  'src',
  'app',
  'pages',
  'components',
  'src/components',
  'src/pages',
  'src/app',
  'src/routes',
  'src/router',
  'src/store',
  'src/stores',
  'src/services',
  'src/api',
  'src/lib',
  'public'
];

const PACKAGE_LOCKS = [
  ['pnpm', 'pnpm-lock.yaml'],
  ['yarn', 'yarn.lock'],
  ['npm', 'package-lock.json'],
  ['bun', 'bun.lockb']
];

const SCAN_IGNORE_DIRS = new Set(['.git', 'node_modules', 'dist', 'build', '.next', 'coverage', '__pycache__']);
const DATA_HINT_TOKENS = ['api', 'service', 'services', 'data', 'store', 'mock', 'fixture', 'country', 'countries'];
const DATA_HINT_EXTENSIONS = new Set(['.js', '.jsx', '.ts', '.tsx', '.mjs', '.cjs', '.json']);
const MAX_DATA_HINTS = 40;
const MAX_SCAN_DEPTH = 6;

function readPackageJson(root) {
  const path = join(root, 'package.json');
  if (!existsSync(path)) return {};
  try {
    return JSON.parse(readFileSync(path, 'utf8'));
  } catch {
    return {};
  }
}

function objectValue(value) {
  return value && typeof value === 'object' && !Array.isArray(value) ? value : {};
}

function deps(packageJson) {
  return {
    ...objectValue(packageJson.dependencies),
    ...objectValue(packageJson.devDependencies)
  };
}

function detectFramework(dependencies) {
  if (dependencies.next) return 'Next.js';
  if (dependencies.react) return 'React';
  if (dependencies.vue) return 'Vue';
  if (dependencies.svelte || dependencies['@sveltejs/kit']) return 'Svelte';
  if (dependencies['@angular/core']) return 'Angular';
  return 'Unknown';
}

function detectBundler(root, dependencies) {
  if (dependencies.vite || existsSync(join(root, 'vite.config.ts')) || existsSync(join(root, 'vite.config.js'))) return 'Vite';
  if (dependencies.next) return 'Next.js';
  if (dependencies.webpack || existsSync(join(root, 'webpack.config.js')) || existsSync(join(root, 'webpack.config.ts'))) return 'Webpack';
  if (dependencies.parcel) return 'Parcel';
  return 'Unknown';
}

function existingCandidates(root, candidates) {
  return candidates.filter((file) => existsSync(join(root, file)));
}

function detectPackageManager(root) {
  const match = PACKAGE_LOCKS.find(([, lockfile]) => existsSync(join(root, lockfile)));
  return match ? `${match[0]} (${match[1]})` : 'Unknown';
}

function extensionOf(file) {
  const index = file.lastIndexOf('.');
  return index === -1 ? '' : file.slice(index);
}

function collectDataHints(root) {
  const hints = [];
  const walk = (dir, relativeDir = '', depth = 0) => {
    if (hints.length >= MAX_DATA_HINTS || depth > MAX_SCAN_DEPTH) return;

    let entries;
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }

    for (const entry of entries.sort((a, b) => a.name.localeCompare(b.name))) {
      if (hints.length >= MAX_DATA_HINTS) return;

      const relative = relativeDir ? `${relativeDir}/${entry.name}` : entry.name;
      if (entry.isDirectory()) {
        if (!SCAN_IGNORE_DIRS.has(entry.name)) {
          walk(join(dir, entry.name), relative, depth + 1);
        }
        continue;
      }

      if (!entry.isFile()) continue;
      const lower = relative.toLowerCase();
      if (!DATA_HINT_EXTENSIONS.has(extensionOf(lower))) continue;
      if (!DATA_HINT_TOKENS.some((token) => lower.includes(token))) continue;
      hints.push(relative);
    }
  };

  walk(root);
  return hints;
}

export function scanFrontendProject(root) {
  const base = resolve(root);
  const packageJson = readPackageJson(base);
  const dependencies = deps(packageJson);
  const scripts = objectValue(packageJson.scripts);
  const entryFiles = existingCandidates(base, ENTRY_FILES);
  const configFiles = existingCandidates(base, CONFIG_FILES);
  const keyDirs = existingCandidates(base, KEY_DIRS);
  const dataHints = collectDataHints(base);

  const lines = [
    '# Frontend Project Scan',
    `Root: ${base}`,
    `Framework: ${detectFramework(dependencies)}`,
    `Bundler: ${detectBundler(base, dependencies)}`,
    `Package Manager: ${detectPackageManager(base)}`,
    '',
    '## Scripts'
  ];

  const scriptEntries = Object.entries(scripts).sort(([a], [b]) => a.localeCompare(b));
  if (scriptEntries.length) {
    lines.push(...scriptEntries.map(([name, cmd]) => `- ${name}: ${cmd}`));
  } else {
    lines.push('- (no package.json scripts found)');
  }

  lines.push('', '## Entry Files');
  if (entryFiles.length) {
    lines.push(...entryFiles.map((file) => `- ${file}`));
  } else {
    lines.push('- (no common frontend entry files found)');
  }

  lines.push('', '## Config Files');
  if (configFiles.length) {
    lines.push(...configFiles.map((file) => `- ${file}`));
  } else {
    lines.push('- (no common frontend config files found)');
  }

  lines.push('', '## Key Directories');
  if (keyDirs.length) {
    lines.push(...keyDirs.map((file) => `- ${file}`));
  } else {
    lines.push('- (no common frontend directories found)');
  }

  lines.push('', '## Data/Service Candidates');
  if (dataHints.length) {
    lines.push(...dataHints.map((file) => `- ${file}`));
  } else {
    lines.push('- (no obvious data, API, service, store, mock, or country files found)');
  }

  lines.push('', '## Dependencies');
  const interesting = Object.keys(dependencies)
    .sort()
    .filter((name) => ['react', 'vue', 'next', 'vite', 'webpack', 'svelte', '@sveltejs/kit', '@angular/core', 'typescript'].includes(name));
  if (interesting.length) {
    lines.push(...interesting.slice(0, 20).map((name) => `- ${name}: ${dependencies[name]}`));
  } else {
    lines.push('- (no frontend dependencies found)');
  }

  return lines.join('\n');
}

File: cc_devtools/project_scan.py
import json
import os
from pathlib import Path


ENTRY_FILES = [
    "src/App.tsx",
    "src/App.jsx",
    "src/App.vue",
    "src/main.tsx",
    "src/main.jsx",
    "src/main.ts",
    "src/main.js",
    "src/pages/index.tsx",
    "src/app/page.tsx",
    "app/page.tsx",
    "pages/index.tsx",
    "src/index.tsx",
    "src/index.jsx",
    "index.html",
]

CONFIG_FILES = [
    "vite.config.ts",
    "vite.config.js",
    "next.config.js",
    "next.config.mjs",
    "webpack.config.js",
    "webpack.config.ts",
    "tsconfig.json",
    "jsconfig.json",
    "tailwind.config.js",
    "postcss.config.js",
]

KEY_DIRS = [
    "src",
    "app",
    "pages",
    "components",
    "src/components",
    "src/pages",
    "src/app",
    "src/routes",
    "src/router",
    "src/store",
    "src/stores",
    "src/services",
    "src/api",
    "src/lib",
    "public",
]

PACKAGE_LOCKS = [
    ("pnpm", "pnpm-lock.yaml"),
    ("yarn", "yarn.lock"),
    ("npm", "package-lock.json"),
    ("bun", "bun.lockb"),
]

SCAN_IGNORE_DIRS = {".git", "node_modules", "dist", "build", ".next", "coverage", "__pycache__"}
DATA_HINT_TOKENS = ("api", "service", "services", "data", "store", "mock", "fixture", "country", "countries")
DATA_HINT_EXTENSIONS = {".js", ".jsx", ".ts", ".tsx", ".mjs", ".cjs", ".json"}
MAX_DATA_HINTS = 40
MAX_SCAN_DEPTH = 6


def _read_package_json(root):
    path = Path(root) / "package.json"
    if not path.exists():
        return {}
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError:
        return {}


def _deps(package):
    merged = {}
    for key in ("dependencies", "devDependencies"):
        value = package.get(key)
        if isinstance(value, dict):
            merged.update(value)
    return merged


def _existing_candidates(base, candidates):
    return [name for name in candidates if (base / name).exists()]


def _detect_package_manager(base):
    for name, lockfile in PACKAGE_LOCKS:
        if (base / lockfile).exists():
            return f"{name} ({lockfile})"
    return "Unknown"


def _relative(path, base):
    return path.relative_to(base).as_posix()


def _collect_data_hints(base):
    hints = []
    for root, dirs, files in os.walk(base):
        current = Path(root)
        rel_parts = current.relative_to(base).parts
        if len(rel_parts) > MAX_SCAN_DEPTH:
            dirs[:] = []
            continue

        dirs[:] = sorted(name for name in dirs if name not in SCAN_IGNORE_DIRS)
        for filename in sorted(files):
            path = current / filename
            rel = _relative(path, base)
            lower = rel.lower()
            if path.suffix.lower() not in DATA_HINT_EXTENSIONS:
                continue
            if not any(token in lower for token in DATA_HINT_TOKENS):
                continue
            hints.append(rel)
            if len(hints) >= MAX_DATA_HINTS:
                return hints
    return hints


def _detect_framework(deps):
    if "next" in deps:
        return "Next.js"
    if "react" in deps:
        return "React"
    if "vue" in deps:
        return "Vue"
    if "svelte" in deps or "@sveltejs/kit" in deps:
        return "Svelte"
    if "@angular/core" in deps:
        return "Angular"
    return "Unknown"


def _detect_bundler(root, deps):
    base = Path(root)
    if "vite" in deps or any((base / name).exists() for name in ("vite.config.ts", "vite.config.js")):
        return "Vite"
    if "next" in deps:
        return "Next.js"
    if "webpack" in deps or any((base / name).exists() for name in ("webpack.config.js", "webpack.config.ts")):
        return "Webpack"
    if "parcel" in deps:
        return "Parcel"
    return "Unknown"


def scan_frontend_project(root):
    base = Path(root).resolve()
    package = _read_package_json(base)
    deps = _deps(package)
    scripts = package.get("scripts") if isinstance(package.get("scripts"), dict) else {}
    existing_entries = _existing_candidates(base, ENTRY_FILES)
    existing_configs = _existing_candidates(base, CONFIG_FILES)
    existing_dirs = _existing_candidates(base, KEY_DIRS)
    data_hints = _collect_data_hints(base)

    lines = [
        "# Frontend Project Scan",
        f"Root: {base}",
        f"Framework: {_detect_framework(deps)}",
        f"Bundler: {_detect_bundler(base, deps)}",
        f"Package Manager: {_detect_package_manager(base)}",
        "",
        "## Scripts",
    ]

    if scripts:
        lines.extend(f"- {name}: {cmd}" for name, cmd in sorted(scripts.items()))
    else:
        lines.append("- (no package.json scripts found)")

    lines.extend(["", "## Entry Files"])
    if existing_entries:
        lines.extend(f"- {name}" for name in existing_entries)
    else:
        lines.append("- (no common frontend entry files found)")

    lines.extend(["", "## Config Files"])
    if existing_configs:
        lines.extend(f"- {name}" for name in existing_configs)
    else:
        lines.append("- (no common frontend config files found)")

    lines.extend(["", "## Key Directories"])
    if existing_dirs:
        lines.extend(f"- {name}" for name in existing_dirs)
    else:
        lines.append("- (no common frontend directories found)")

    lines.extend(["", "## Data/Service Candidates"])
    if data_hints:
        lines.extend(f"- {name}" for name in data_hints)
    else:
        lines.append("- (no obvious data, API, service, store, mock, or country files found)")

    lines.extend(["", "## Dependencies"])
    if deps:
        interesting = [name for name in sorted(deps) if name in {"react", "vue", "next", "vite", "webpack", "svelte", "@sveltejs/kit", "@angular/core", "typescript"}]
        lines.extend(f"- {name}: {deps[name]}" for name in interesting[:20])
    else:
        lines.append("- (no frontend dependencies found)")

    return "\n".join(lines)
